cmd/kubeadm/app/util/config/common.go

```diff
@@ -181,3 +160,46 @@ func ChooseAPIServerBindAddress(bindAddress net.IP) (net.IP, error) {
 	}
 	return ip, nil
 }
+
+// MigrateOldConfigFromFile migrates an old configuration from a file into a new one (returned as a byte slice). Only kubeadm kinds are migrated. Others are silently ignored.
+func MigrateOldConfigFromFile(cfgPath string) ([]byte, error) {
+	newConfig := [][]byte{}
+
+	cfgBytes, err := ioutil.ReadFile(cfgPath)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	gvks, err := kubeadmutil.GroupVersionKindsFromBytes(cfgBytes)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Migrate InitConfiguration and ClusterConfiguration if there are any in the config
+	if kubeadmutil.GroupVersionKindsHasInitConfiguration(gvks...) || kubeadmutil.GroupVersionKindsHasClusterConfiguration(gvks...) {
+		o, err := ConfigFileAndDefaultsToInternalConfig(cfgPath, &kubeadmapiv1beta1.InitConfiguration{})
```

Comment:
I think we can do some refactoring for `ConfigFileAndDefaultsToInternalConfig`, since the file will be read twice. There is no need.

Reply:
Yes, that's on my agenda for the next release cycle. Decoupling InitConfiguration and ClusterConfiguration is one of my priorities. If we refactor the APIs and make ClusterConfiguration more standalone and introducing an internal type that houses the usual pair (ClusterConfiguration + NodeRegistrationOptions for the current node) will eliminate most of the confusion and optimize things at all levels.
But for now at the end of this release cycle it's best to keep the changes as small as possible and just fix the bug at hand.